components/camera/videosource/query.go

```diff
@@ -0,0 +1,225 @@
+	filter driver.FilterFn,
+	constraints mediadevices.MediaTrackConstraints,
+	logger logging.Logger,
+) (driver.Driver, prop.Media, error) {
+	var bestDriver driver.Driver
+	var bestProp prop.Media
+	minFitnessDist := math.Inf(1)
+
+	baseDrivers := driver.GetManager().Query(baseFilter)
+	logger.Debugw("before specific filter, we found the following drivers", "count", len(baseDrivers))
+	for _, d := range baseDrivers {
+		logger.Debugw(d.Info().Label, "priority", float32(d.Info().Priority), "type", d.Info().DeviceType)
+	}
+
+	driverProperties := queryDriverProperties(filter, logger)
+	if len(driverProperties) == 0 {
+		logger.Debugw("found no drivers matching filter")
+	} else {
+		logger.Debugw("found drivers matching specific filter", "count", len(driverProperties))
+	}
+	for d, props := range driverProperties {
+		priority := float64(d.Info().Priority)
+		logger.Debugw(
+			"considering driver",
+			"label", d.Info().Label,
+			"priority", priority)
+		for _, p := range props {
+			fitnessDist, ok := constraints.MediaConstraints.FitnessDistance(p)
+			if !ok {
+				logger.Debugw("driver does not satisfy any constraints", "label", d.Info().Label)
+				continue
+			}
+			fitnessDistWithPriority := fitnessDist - priority
+			logger.Debugw(
+				"driver properties satisfy some constraints",
+				"label", d.Info().Label,
+				"props", p,
+				"distance", fitnessDist,
+				"distance_with_priority", fitnessDistWithPriority)
+			if fitnessDistWithPriority < minFitnessDist {
+				minFitnessDist = fitnessDistWithPriority
+				bestDriver = d
+				bestProp = p
+			}
+		}
+	}
+
+	if bestDriver == nil {
+		return nil, prop.Media{}, errors.New("failed to find the best driver that fits the constraints")
+	}
+
+	logger.Debugw("winning driver", "label", bestDriver.Info().Label, "props", bestProp)
+	selectedMedia := prop.Media{}
+	selectedMedia.MergeConstraints(constraints.MediaConstraints)
+	selectedMedia.Merge(bestProp)
+	return bestDriver, selectedMedia, nil
+}
+
+func queryDriverProperties(
+	filter driver.FilterFn,
+	logger logging.Logger,
+) map[driver.Driver][]prop.Media {
+	var needToClose []driver.Driver
+	drivers := driver.GetManager().Query(filter)
+	m := make(map[driver.Driver][]prop.Media)
+
+	for _, d := range drivers {
+		var status string
+		isAvailable, err := driver.IsAvailable(d)
+		if errors.Is(err, availability.ErrUnimplemented) {
+			s := d.Status()
+			status = string(s)
+			isAvailable = s == driver.StateClosed
+		} else if err != nil {
+			status = err.Error()
+		}
+
+		if isAvailable {
+			err := d.Open()
+			if err != nil {
+				logger.Debugw("error opening driver for querying", "error", err)
+				// Skip this driver if we failed to open because we can't get the properties
+				continue
+			}
+			needToClose = append(needToClose, d)
+			m[d] = d.Properties()
+		} else {
+			logger.Debugw("driver not available", "name", d.Info().Name, "label", d.Info().Label, "status", status)
+		}
+	}
+
+	for _, d := range needToClose {
```

Comment:
Should we just close the driver after getting Properties in the main loop?

Does not seem like that big of a deal if we need to open everything then close out afterwards as is.

Reply:
This is a reasonable suggestion... but I'm a bit hesitant touching this code since it's stuff copy and pasted from gostream. Feels a bit out of scope and could cause confusion since there's no git history of it being here. WDYT @randhid

Reply:
Agree with Sean Yu, we can address more code refactoring/optimizations for webcam when we have time.

Reply:
Sounds good